storage/remote/queue_manager.go

```diff
@@ -844,38 +846,70 @@ func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.Ti
 		return err
 	}
 
+	try := 0
+
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
 		default:
 		}
-		begin := time.Now()
-		err := s.qm.client().Store(ctx, req)
-
-		s.qm.metrics.sentBatchDuration.Observe(time.Since(begin).Seconds())
 
+		err = s.attemptSendSamples(ctx, samples, buf, highest, try)
 		if err == nil {
-			s.qm.metrics.succeededSamplesTotal.Add(float64(len(samples)))
-			s.qm.metrics.bytesSent.Add(float64(len(req)))
-			s.qm.metrics.highestSentTimestamp.Set(float64(highest / 1000))
 			return nil
 		}
 
+		// If we make it this far, we've encountered an unrecoverable error and should not retry.
 		if _, ok := err.(recoverableError); !ok {
 			return err
 		}
-		s.qm.metrics.retriedSamplesTotal.Add(float64(len(samples)))
-		level.Warn(s.qm.logger).Log("msg", "Failed to send batch, retrying", "err", err)
 
+		// If we make it this far, we've encountered a recoverable error and will retry.
 		time.Sleep(time.Duration(backoff))
 		backoff = backoff * 2
 		if backoff > s.qm.cfg.MaxBackoff {
 			backoff = s.qm.cfg.MaxBackoff
 		}
+
+		try++
 	}
 }
 
+func (s *shards) attemptSendSamples(ctx context.Context, samples []prompb.TimeSeries, req *[]byte, highest int64, try int) error {
+	reqSize := len(*req)
+	sampleCount := len(samples)
+
+	span, ctx := opentracing.StartSpanFromContext(ctx, "Remote Send Batch")
+	defer span.Finish()
+
+	span.SetTag("samples", sampleCount)
+	span.SetTag("requestSize", reqSize)
+	span.SetTag("try", try)
+	span.SetTag("remote_name", s.qm.storeClient.Name())
+	span.SetTag("remote_url", s.qm.storeClient.Endpoint())
+
+	begin := time.Now()
+	err := s.qm.client().Store(ctx, *req)
+
+	s.qm.metrics.sentBatchDuration.Observe(time.Since(begin).Seconds())
+
+	if err == nil {
```

Comment:
Now that this is in a different function, could we invert the logic to the more normal `if err != nil` flow?

Reply:
Yea, copy/paste got away from me there. :-D